### PR TITLE
Use enums/booleans for enumerated attributes.

### PR DESCRIPTION
**Pull request by aptls**
Modify the attribute getters/setters for attributes defined in
the DTD with enumerated values. Use enums for most string values
and boolean where the values in the DTD are "true|false".

The behavior on the getters is currently to return the value of
the member variable, returning null when the value is unset
(false for boolean attributes). This is different from the
generated methods, which returned the default attribute value
as specified in the DTD even when the member variable was unset. The intent
is to model what will be present in the generated XML.

Add tests that verify that the expected enum/boolean is returned
on unmarshal and that the setters result in the correct string in
the XML.

### CODE REVIEW

File: src/com/google/enterprise/gsafeed/Acl.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Acl.java]: Maybe:

    for (InheritanceType type : InheritanceType.values) {
      if (type.xmlValue.equals(value)) {
        return type;
      }
    }
    throw new IllegalArgumentException(value);

File: src/com/google/enterprise/gsafeed/Acl.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Acl.java]: This seems awkward. Can we marshal the enum values directly, rather than store them as strings. This post at least suggests that we can:
http://blog.bdoughan.com/2011/08/jaxb-and-enums.html

File: src/com/google/enterprise/gsafeed/Acl.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Acl.java]: What's the purpose of allowing empty strings to act like nulls?

Maybe, if we don't allow empty strings, and we keep the string representation of enums, we could move the null check into fromString and remove it from the getter and String setter.

File: src/com/google/enterprise/gsafeed/Metadata.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Metadata.java]: This comment seems odd in the getter. What does it mean to not the set string here? Also, you're not allowing setOverwriteAcls(null) to reset the default, like you do for the enums. Maybe at least allow that for setOverwriteAcls(String), and maybe think about using Boolean instead of boolean, and live with the autoboxing?

File: src/com/google/enterprise/gsafeed/Record.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Record.java]: s/Scorng/Scoring/

File: test/com/google/enterprise/gsafeed/AclTest.java
Comment (wiarlawd) [inline test/com/google/enterprise/gsafeed/AclTest.java]: If this exception isn't thrown until the getter is called, then move these down to make that more clear (and not spuriously pass the test if unmarshal throws an IllegalArgumentException). Otherwise, remove the call to the getter.

File: test/com/google/enterprise/gsafeed/AclTest.java
Comment (wiarlawd) [inline test/com/google/enterprise/gsafeed/AclTest.java]: These aren't used. Maybe you added JaxbUtil later on?

File: src/com/google/enterprise/gsafeed/Acl.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Acl.java]: If we keep this string representation, this code block is pretty repetitive. Maybe:

    this.inheritanceType = (value == null) ? null : value.toString();
or

    this.inheritanceType = toString(value, null);

where toString is an implementation of Objects.toString (which we can't use; it's Java 7).

**Review by wiarlawd — COMMENTED** [review]
Minor stuff, questions, and suggestions.

File: test/com/google/enterprise/gsafeed/AclTest.java
Comment (wiarlawd) [inline test/com/google/enterprise/gsafeed/AclTest.java]: This is the price we pay for unmarshalling to an enum? Invalid values are just silently ignored? What happens with an invalid value if we unmarshal the full XML using the DTD?

File: test/com/google/enterprise/gsafeed/AclTest.java
Comment (wiarlawd) [inline test/com/google/enterprise/gsafeed/AclTest.java]: Maybe add a test for a valid use of fromString?

File: src/com/google/enterprise/gsafeed/Acl.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Acl.java]: In looking at this, we get fromString(null) -> IllegalArgumentException(null). Do we want either fromString(null) -> null (since that's a valid argument to the setters, and it avoids callers who want to use fromString from making the check themselves), or fromString(null) -> NPE?

**Review by wiarlawd — COMMENTED** [review]
This looks really clean.

File: test/com/google/enterprise/gsafeed/AclTest.java
Comment (aptls) [reply test/com/google/enterprise/gsafeed/AclTest.java]: This is what we get without validation. With validation, the parser throws an exception.

File: src/com/google/enterprise/gsafeed/Content.java
Comment (wiarlawd) [inline src/com/google/enterprise/gsafeed/Content.java]: Base64 seems like one word to me, so I think these (and the one for meta-encoding) should be BASE64.... I also wonder about having a fixed rule for the strings, like the value.replace("-", "_").toUpperCase(US) we talked about, so that this would be BASE64BINARY by virtue of the unhyphenated string value. Not convinced either way.

**Review by wiarlawd — APPROVED** [review]
LGTM++ with a waffle.

Please wait for PJ's approval as well.

File: build.xml
Comment (PapaJoltOmega) [inline build.xml]: where are these coming from? they're not in this CL?

**Review by PapaJoltOmega — APPROVED** [review]
Thank you

File: build.xml
Comment (wiarlawd) [reply build.xml]: Took me a little while to figure that out as well. The "tests", such as they were, did not exercise enough prod code to hit the prod dependency on Guava. We could instead include gsa.build.classpath here, which might be slightly more conventional.

Comment (wiarlawd) [reply build.xml]: I was mistaken about this. The initial commit was using Strings.isNullOrEmpty, but that's been removed, and this can go away. The only use of Guava in gsafeed is an `@VisibleForTesting` in the copied GsaFeedFileSender.

File: src/com/google/enterprise/gsafeed/Content.java
Comment (wiarlawd) [reply src/com/google/enterprise/gsafeed/Content.java]: No one else seems to have an opinion, so just s/BASE_64/BASE64/g.

**Review by wiarlawd — APPROVED** [review]
LGTM++, resolving issues.

**Pull request merged**
